File: OJ/usaco/cpp/nocows.cpp
/*
ID:oifox20071
LANG:C++
PROG:nocows
*/
#include <iostream>
#include <cstring>
using namespace std;
const int prime = 9901;
const int maxn = 200;
const int maxm = 100;
int f[maxn][maxm];
int main()
{
	freopen("nocows.in", "r", stdin); freopen("nocows.out", "w", stdout);
	int n, m;
	cin >> n >> m;
	memset(f, 0, sizeof(f));
	for (int i = 1; i <= m; ++ i)
		f[1][i] = 1;
	for (int i = 2; i <= n; ++ i)
		for (int j = 1; j <= m; ++ j)
			for (int k = 1; k <= i - 2; ++ k)
				f[i][j] = (f[i][j] + (f[k][j - 1] * f[i - k - 1][j - 1]) % prime) % prime;
	int ans = (f[n][m] - f[n][m - 1] + prime) % prime;
	cout << ans << endl;
	return 0;
}

File: OJ/usaco/pascal/s2/s23/nocows.pas
{
ID:lijianx1
PROG:nocows
LANG:PASCAL
}
program nocows;
const maxn=200;
      maxk=100;
      prime=9901;
var f:array[1..maxn,1..maxk] of longint;
    n,m:longint;
procedure init;
 begin
     assign(input,'nocows.in'); reset(input);
     readln(n,m);
     close(input);
 end;

procedure main;
 var i,j,k:longint;
 begin
     fillchar(f,sizeof(f),0);
     for i:=1 to m do f[1,i]:=1;
     for i:=2 to n do
	for j:=1 to m do 
	   for k:=1 to i-2 do
		f[i,j]:=(f[i,j]+(f[k,j-1]*f[i-1-k,j-1]) mod prime) mod prime;
 end;

procedure print;
 var ans:longint;
 begin
     ans:=(f[n,m]-f[n,m-1]+prime) mod prime;
     assign(output,'nocows.out'); rewrite(output);
     writeln(ans);
     close(output);
 end;

begin
    init;
    main;
    print;
end.
